components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.role.mgt.core;
+
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.context.CarbonContext;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.role.mgt.core.dao.RoleDAO;
+import org.wso2.carbon.identity.role.mgt.core.dao.RoleDAOImpl;
+import org.wso2.carbon.user.core.UserRealm;
+import org.wso2.carbon.user.mgt.UserRealmProxy;
+import org.wso2.carbon.user.mgt.common.UIPermissionNode;
+import org.wso2.carbon.user.mgt.common.UserAdminException;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.role.mgt.core.RoleConstants.Error.UNEXPECTED_SERVER_ERROR;
+
+/**
+ * Implementation of the {@link RoleManagementService} interface.
+ */
+public class RoleManagementServiceImpl implements RoleManagementService {
+
+    private static final Log log = LogFactory.getLog(RoleManagementServiceImpl.class);
+
+    @Override
+    public RoleBasicInfo addRole(String roleName, List<String> userList, List<String> groupList,
+            List<String> permissions, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
```

Comment:
Has this been the model we followed. Shouldn't the dao be a composite of the implementation and instantiated at the constructor.

Reply:
Updated the PR with the new changes.